class-b2c-token-checker.php

```diff
@@ -0,0 +1,151 @@
+
+/** 
+ * A class to verify an id_token, following the implicit flow
+ * defined by the OpenID Connect standard.
+ */
+class B2C_Token_Checker {
+	
+	private $id_token_array = array(); // still encoded
+	private $head = array(); // decoded
+	private $payload = array(); // decoded
+	private $clientID = '';
+	private $endpoint_handler; 
+	
+	function __construct($id_token, $clientID, $policy_name) {
+		
+		$this->clientID = $clientID;
+		$this->endpoint_handler = new B2C_Endpoint_Handler($policy_name);
+		$this->split_id_token($id_token);
+	}
+	
+	/** 
+	 * Converts base64url encoded string into base64 encoded string.
+	 * Also adds the necessary padding to the base64 encoded string.
+	 */
+	private function convert_base64url_to_base64($data) {
+		return str_pad(strtr($data, '-_', '+/'), strlen($data) % 4, '=', STR_PAD_RIGHT);
+	}
+	
+	/**
+	 * Splits the id token into an array of header, payload, and signature.
+	 */
+	private function split_id_token($id_token) {
+
+		// Split the token into Header, Payload, and Signature, and decode
+		$this->id_token_array = explode('.', $id_token);
+		$this->head = json_decode(base64_decode($this->id_token_array[0]), true);
+		$this->payload = json_decode(base64_decode($this->id_token_array[1]), true);
+	}
+	
+	/** 
+	 * Validates the RSA signature on the token.
+	 */
+	private function validate_signature() {
+		
+		//require_once "phpseclib/Crypt/RSA.php";
+		
+		// Get kid from header
+		$kid = $this->head['kid'];
+		
+		// For each JwksURI, get the public key and verify the signature
+		$key_datas = $this->endpoint_handler->get_jwks_uri_data();
+		foreach ($key_datas as $key_data) {
+			
+			// Iterate through each key type until the one that matches the "kid" is found
+			$keys = json_decode($key_data, true)['keys'];
+			foreach ($keys as $key) {
+				
+				if ($key['kid'] == $kid) {
+					$e = $key['e'];
+					$n = $key['n'];
+					break; 
+				}
+			}
+			
+			// 'e' and 'n' are base64 URL encoded, change to just base64 encoding
+			$e = $this->convert_base64url_to_base64($e);
+			$n = $this->convert_base64url_to_base64($n);
+
+			// Convert RSA(e,n) format to PEM format
+			$rsa = new Crypt_RSA();
+			$rsa->setPublicKey('<RSAKeyValue>
+				<Modulus>' . $n . '</Modulus>
+				<Exponent>' . $e . '</Exponent>
+				</RSAKeyValue>');
+			$public_key = $rsa->getPublicKey();
+			
+			// Construct data and signature for verification
+			$to_verify_data = $this->id_token_array[0] . "." . $this->id_token_array[1];
+			$to_verify_sig = base64_decode($this->convert_base64url_to_base64(($this->id_token_array[2])));
+			
+			try {
+				$jwt = $this->id_token_array[0] . "." . $this->id_token_array[1] . "." .$this->id_token_array[2];
+				$decoded = JWT::decode($jwt, $public_key, array('HS256', 'HS384', 'HS512', 'RS256'));
+			}
+			catch (Exception $e) {
+				echo 'error: ' . $e->getMessage();
+				return false;
+			}
+			
+		}
+		
+		// Returns true when verified successfully
+		return true; 
+	}
```

Comment:
This still needs to be cleaned up a bit.

`$this->endpoint_handler->get_jwks_uri_data()` should be returning a *single* JWK Set (a *single* set of one or more JWKs), since the OpenID Connect metadata `jwks_uri` field will only return one value. Why is this implemented as if there could be an array of `jwks_uri` values, which would lead to an array of JWK Sets?

Splitting and decoding the JWT is not necessary. The only think you need to parse is the JWK Set:

1. Download and JSON-decode the (single) JWK Set. This is a single JSON object that has a `keys` member, which is an array of JWKs. Let's say the JWK Set is `$jwks`, and so the array of JWKs is in `$jwks->keys`.
2. Iterate over each JWK, picking out valid candidates (i.e. those with a `kty` that you support, and those with a `use` set to signing (`sig`)). Create a new associative array (let's say it's `$signing_keys`) where each element's key is the candidate JWK's `kid` value, and the element's value is the candidate JWK's PEM-formatted key value.
3. Pass the original JWT that you received, the array of candidate signing keys, and the list of supported and expected signing algorithms to `JWT::decode`.

Here's an untested example of what I mean:

````php
$supported_algs = array('HS256', 'HS384', 'HS512', 'RS256');

// Get the *single* JWK Set from the 'jwks_uri' endpoint
$jwks = $this->endpoint_handler->get_jwks_uri_data();
    
// Construct a list of candidate signing keys ('kid' => 'PEM-formatted key')
$signing_keys = array();
foreach ( $jwks->keys as $jwk ) {
  if ( 'rsa' == $jwk->kty && 'sig' == $jwk->use ) {
    $signing_keys[ $jwk->kid ] = $this->get_pem_from_n_e( $jwk->n, $jwk->e );
  }
}

try {
  $decoded = JWT::decode($jwt, $signing_keys, $supported_algs);
} catch ( Exception $e ) {
  echo 'Error decoding ID Token: ' . $e->getMessage();
  return false;
}
````